src/sync/locks.rs

```diff
@@ -0,0 +1,198 @@
+  raw: RawMutex,
+  data: UnsafeCell<T>,
+}
+impl<T> Mutex<T> {
+  /// Creates a new lock containing a given value.
+  pub const fn new(t: T) -> Self {
+    Mutex { raw: RawMutex::new(), data: UnsafeCell::new(t) }
+  }
+
+  /// Returns a guard for this lock, or panics if there is another lock active.
+  pub fn lock(&self) -> MutexGuard<'_, T> {
+    self.try_lock().unwrap_or_else(|| already_locked())
+  }
+
+  /// Returns a guard for this lock or `None` if there is another lock active.
+  pub fn try_lock(&self) -> Option<MutexGuard<'_, T>> {
+    if self.raw.raw_lock() {
+      Some(MutexGuard { underlying: self, ptr: self.data.get() })
+    } else {
+      None
+    }
+  }
+}
+unsafe impl<T> Send for Mutex<T> {}
+unsafe impl<T> Sync for Mutex<T> {}
+
+/// A guard representing an active lock on an [`Mutex`].
+pub struct MutexGuard<'a, T> {
+  underlying: &'a Mutex<T>,
+  ptr: *mut T,
+}
+impl<'a, T> Drop for MutexGuard<'a, T> {
+  fn drop(&mut self) {
+    self.underlying.raw.raw_unlock();
+  }
+}
+impl<'a, T> Deref for MutexGuard<'a, T> {
+  type Target = T;
+  fn deref(&self) -> &Self::Target {
+    unsafe { &*self.ptr }
+  }
+}
+impl<'a, T> DerefMut for MutexGuard<'a, T> {
+  fn deref_mut(&mut self) -> &mut Self::Target {
+    unsafe { &mut *self.ptr }
+  }
+}
+
+enum Void {}
+
+/// A helper type that ensures a particular value is only initialized once.
+pub struct InitOnce<T> {
+  state: Static<u8>,
+  value: UnsafeCell<MaybeUninit<T>>,
+}
+impl<T> InitOnce<T> {
+  /// Creates a new uninitialized object.
+  pub const fn new() -> Self {
+    InitOnce { state: Static::new(0), value: UnsafeCell::new(MaybeUninit::uninit()) }
+  }
+
+  /// Gets the contents of this state, or initializes it if it has not already
+  /// been initialized.
+  ///
+  /// The initializer function is guaranteed to only be called once.
+  ///
+  /// Take care when sharing an `InitOnce` object between an IRQ and normal
+  /// code. If this function is called in an IRQ when it is already currently
+  /// being initialized by user code, this function will panic.
+  pub fn get(&self, initializer: impl FnOnce() -> T) -> &T {
+    match self.try_get(|| -> Result<T, Void> { Ok(initializer()) }) {
+      Ok(v) => v,
+      _ => unimplemented!(),
+    }
+  }
+
+  /// Gets the contents of this state, or initializes it if it has not already
+  /// been initialized.
+  ///
+  /// The initializer function is guaranteed to only be called once if it
+  /// returns `Ok`. If it returns `Err`, it will be called again in the
+  /// future until an attempt at initialization succeeds.
+  ///
+  /// Take care when sharing an `InitOnce` object between an IRQ and normal
+  /// code. If this function is called in an IRQ when it is already currently
+  /// being initialized by user code, this function will panic.
+  pub fn try_get<E>(&self, initializer: impl FnOnce() -> Result<T, E>) -> Result<&T, E> {
+    unsafe {
+      if self.state.read() != 2 {
+        // Locks the initializer
```

Comment:
There's a race between the read not being 2 and the replace(1).

Reply:
There is, but the only way this race can end is a panic. As far as I can tell, there's basically two possible sequences of events here:

* The main thread checks `self.state.read() != 2`
* An IRQ fires -- interrupting the main thread entirely.
* The IRQ checks `self.state.read() != 2`, and it passes, then it checks `self.state.replace(1) != 0` which also passes.
* The IRQ eventually completes.
* The main thread checks `self.state.replace(1) != 0`, which fails causing a panic.

Or, alternatively:

* The main thread checks `self.state.read() != 2` and `self.state.replace(1) != 0`.
* An IRQ fires -- interrupting the main thread entirely. It checks `self.state.replace(1) != 0` eventually, which causes a panic.

In either case, the only thing we can reasonably do is panic (or return an error -- I might legitimately want to add an API for doing that, now that I think about it.) We have no way of blocking because the only way to leave the IRQ is for the interrupt handler to return normally, and that isn't going to happen if it's waiting on an lock.

Am I missing something here?

(Note that an IRQ can't trigger in an IRQ because the handler `gba` provides disables interrupts during the handler.)

Reply:
... nvm, I realized something I could be doing about this.